app/org/sagebionetworks/bridge/upload/IosSchemaValidationHandler.java

```diff
@@ -0,0 +1,559 @@
+                    // We expect exactly one field (since we don't know how to fill in the other fields). But if
+                    // there's more than one (as long as one of them is a ATTACHMENT_TYPE_SET), we should be okay.
+                    addMessageAndWarn(context, String.format(
+                            "upload ID %s filename %s identified schema ID %s rev %d has multiple fields", uploadId,
+                            filename, schemaId, schemaRev));
+                }
+
+                UploadFieldDefinition attachmentFieldDef = null;
+                for (UploadFieldDefinition oneFieldDef : fieldDefList) {
+                    if (ATTACHMENT_TYPE_SET.contains(oneFieldDef.getType())) {
+                        // just the first one
+                        attachmentFieldDef = oneFieldDef;
+                        break;
+                    }
+                }
+                if (attachmentFieldDef == null) {
+                    // We have no field to write this into. Game over.
+                    throw new UploadValidationException(String.format(
+                            "Identified schema ID %s rev %d has no field for non-JSON data", schemaId, schemaRev));
+                }
+
+                // Write this to the attachment map. UploadArtifactsHandler will take care of the rest.
+                attachmentMap.put(attachmentFieldDef.getName(), fileData);
+            }
+        } else {
+            // This means our data is in JSON format, so we can look inside it to figure out what it is.
+
+            if (!unzippedDataMap.isEmpty()) {
+                // We don't expect a mix of JSON and non-JSON, but we can just ignore the non-JSON data.
+                addMessageAndWarn(context, String.format(
+                         "upload ID %s contains both JSON data and non-JSON data; ignoring non-JSON data", uploadId));
+            }
+
+            // Check if it's a survey. We can tell if it's a survey by looking at one of the JSON files (other than
+            // info.json) and looking for specific survey keys, as listed in SURVEY_KEY_SET.
+            boolean isSurvey = false;
+            for (Map.Entry<String, JsonNode> oneJsonFile : jsonDataMap.entrySet()) {
+                if (oneJsonFile.getKey().equals("info.json")) {
+                    // Not info.json. Skip.
+                    continue;
+                }
+
+                // There may be fields beyond what's specified in SURVEY_KEY_SET. This is normal, since every question
+                // type has its own special fields.
+                JsonNode oneJsonFileNode = oneJsonFile.getValue();
+                Set<String> fieldNameSet = ImmutableSet.copyOf(oneJsonFileNode.fieldNames());
+                isSurvey = fieldNameSet.containsAll(SURVEY_KEY_SET);
+
+                // We only need to look at one (other than info.json). Either they're all surveys, or none of them are.
+                break;
+            }
+
+            if (isSurvey) {
```

Comment:
This has been refactored into bite-sized helper functions. I've organized it in a way that hopefully makes it more readable. At some point, I'll draw a diagram that explains how everything fits together.